### PR TITLE
feat(twitter): Add predictive back gesture patch

### PR DESCRIPTION
**Pull request by bennett-sh**
This patch enables the new Android 13 predictive back gesture for Twitter.

### CODE REVIEW

File: src/main/kotlin/app/revanced/patches/twitter/misc/predictiveback/patch/PredictiveBackPatch.kt [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+import app.revanced.patcher.patch.PatchResult
+import app.revanced.patcher.patch.PatchResultSuccess
+import app.revanced.patcher.patch.ResourcePatch
+import app.revanced.patcher.patch.annotations.Patch
+import app.revanced.patches.twitter.misc.predictiveback.annotations.PredictiveBackCompatibility
+
+@Patch
+@Name("predictive-back")
+@Description("Enables the new predictive back feature")
+@PredictiveBackCompatibility
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/twitter/misc/predictiveback/patch/PredictiveBackPatch.kt:16]: Does this patch technically not work with every app? If so, you can remove this constraining annotation. The patch would also require to check if the attribute already exists.

File: src/main/kotlin/app/revanced/patches/twitter/misc/predictiveback/patch/PredictiveBackPatch.kt [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+@Version("0.0.1")
+class PredictiveBackPatch : ResourcePatch {
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/twitter/misc/predictiveback/patch/PredictiveBackPatch.kt:18]: Probably `PredictiveBackGesturePatch` would be better here.

**Review by oSumAtrIX — CHANGES_REQUESTED** [review]

File: src/main/kotlin/app/revanced/patches/twitter/misc/predictiveback/patch/PredictiveBackPatch.kt (hunk 1 above)
Comment (bennett-sh) [reply src/main/kotlin/app/revanced/patches/twitter/misc/predictiveback/patch/PredictiveBackPatch.kt:16]: Yes, it does. But in which folder should I move it into?

**Review by oSumAtrIX — COMMENTED** [review]
You forgot to commit the shared directory.

**Review by oSumAtrIX — APPROVED** [review]

**Comment by Vucko130** [comment]
image

**Comment by thegravity98** [comment]
Does it have any effect in lower version of Android 12, 11, ....

**Comment by bennett-sh** [comment]
No it doesn't but to my knowledge there's no way to specify the minimum
Android version in Revanced.

**Comment by kazimmt** [comment]
It should only be limited for ReVanced supported apps.

Otherwise the app list inside ReVanced Manager is biGGGGGGGGGGGGG

Screenshot_2022-12-12-01-34-22-596_app revanced manager flutter
Screenshot_2022-12-12-01-34-03-274_app revanced manager flutter
Screenshot_2022-12-12-01-34-10-964_app revanced manager flutter
Screenshot_2022-12-12-01-34-14-711_app revanced manager flutter
Screenshot_2022-12-12-01-34-19-512_app revanced manager flutter

**Comment by bennett-sh** [comment]
I don't think you should limit this patch but instead the manager should only show apps with non-generic patches so if you want to patch any APK, it will work but the app list won't be filled.
Someone has already created an issue [here](https://github.com/revanced/revanced-manager/issues/579).

**Comment by Canny1913** [comment]
> I don't think you should limit this patch but instead the manager should only show apps with non-generic patches so if you want to patch any APK, it will work but the app list won't be filled. Someone has already created an issue [here](https://github.com/revanced/revanced-manager/issues/579).

We already have a few ideas, such as sorting app list based on most amount of patches or adding an "Load more apps" button.

**Comment by bennett-sh** [comment]
That's probably the best solution.

**Comment by Vucko130** [comment]
how to solve that missing compatibility annotation  problem?

**Comment by Vucko130** [comment]
> This is not a problem.

But builder is skipping that patch because of this problem.

**Comment by oSumAtrIX** [comment]
There is no builder.

**Comment by bennett-sh** [comment]
> how to solve that missing compatibility annotation problem?
it's already fixed

**Comment by Vucko130** [comment]
> > how to solve that missing compatibility annotation problem?
> > it's already fixed

Where? You opened a pr to fix that maybe but it's closed.

File: src/main/kotlin/app/revanced/patches/all/interaction/gestures/patch/PredictiveBackGesturePatch.kt [hunk 3]
```diff
@@ -8,26 +8,29 @@ import app.revanced.patcher.patch.PatchResult
...
-            if(application.attributes.getNamedItem("android:enableOnBackInvokedCallback") == null) {
-                val attr = document.createAttribute("android:enableOnBackInvokedCallback")
-                attr.value = "true"
-                application.attributes.setNamedItem(attr)
+            with(document.getElementsByTagName("application").item(0)) {
+                attributes.getNamedItem(FLAG)?.let {
+                    document.createAttribute(FLAG)
+                        .apply { value = "true" }
+                        .let(attributes::setNamedItem)
+                }
```
Comment (Sculas) [inline src/main/kotlin/app/revanced/patches/all/interaction/gestures/patch/PredictiveBackGesturePatch.kt:26]: @oSumAtrIX This looks very incorrect. The previous code checked if `attributes.getNamedItem("android:enableOnBackInvokedCallback")` is `null` before creating the attribute, but your change instead creates the attribute **only if the attribute already exists** (`?.let` means "if not null, run this block")

Comment (oSumAtrIX) [reply src/main/kotlin/app/revanced/patches/all/interaction/gestures/patch/PredictiveBackGesturePatch.kt:26]: Its not very incorrect, but the logic should be inverted, please open a PR or issue.

**Pull request merged**